### PR TITLE
Implementing PriceFunction interface for Cloud Service Provider sourc…

### PR DESCRIPTION
**Pull request by LSZ-PL**
…ed pricing retrieval

plus Pricing Supplier to actually retrieve full Price Plan using Kafka messaging

### CODE REVIEW

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/OclServiceModule.java [hunk 1]
```diff
@@ -33,7 +35,8 @@ protected void configure() {
     expireBinder.addBinding().to(SolutionCacheImpl.class);
     expireBinder.addBinding().to(NodeCandidateCache.class);
 
-    bind(PriceFunction.class).to(HardwareBasedPriceFunction.class);
+    //bind(PriceFunction.class).to(HardwareBasedPriceFunction.class);
```
Comment (pszkup7bulls) [inline matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/OclServiceModule.java]: to remove.

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/CSPSourcedPricePlanPriceFunction.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                    ,image.getOperatingSystem().getFamily().getName()
+                    ,sha256PriceKey.asLong()));
+            return -1;
+        }
+
+        return getPriceMap(cloud, userId).get(sha256PriceKey.asLong()).getPrice();
+    }
+
+    private Map<Long, IaasEntities.Price> getPriceMap(Cloud cloud, String userId) {
+        if (priceMap == null) {
```
Comment (pszkup7bulls) [inline matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/CSPSourcedPricePlanPriceFunction.java]: Create map with key combined with cloud and userId, and then use computeIfAbsent() method to fetch proper pricing map.

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+            return prices.stream().collect(Collectors.toMap(price -> {
+                HashCode sha256PriceKey = Hashing.sha256()
+                    .hashString((new StringBuilder())
+                            .append(price.getHardwareProviderId())
+                            .append(price.getLocationProviderId())
+                            .append(price.getOsArchitecture())
+                            .append(price.getOsFamily())
+                            .append(price.getCloudAPIProviderName()), StandardCharsets.UTF_8);
+            return sha256PriceKey.asLong();}, price -> price));
```
Comment (pszkup7bulls) [inline matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java:46]: price -> price could be replaced by Function.identity()

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        } catch (ResponseException e) {
+            throw new IllegalStateException(String
+                    .format("Could not retrieve price due to communication error: %s", e.getMessage()), e);
```
Comment (pszkup7bulls) [inline matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java]: "Could not retrieve price due to communication error: " - this is enough

**Review by pszkup7bulls — CHANGES_REQUESTED** [review]

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/OclServiceModule.java (hunk 1 above)
Comment (LSZ-PL) [reply matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/OclServiceModule.java]: done

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/CSPSourcedPricePlanPriceFunction.java (hunk 2 above)
Comment (LSZ-PL) [reply matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/CSPSourcedPricePlanPriceFunction.java]: done

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java (hunk 3 above)
Comment (LSZ-PL) [reply matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java:46]: decided to leave it as is

File: matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java (hunk 4 above)
Comment (LSZ-PL) [reply matchmaking-agent/src/main/java/org/cloudiator/matchmaking/ocl/PricingSupplier.java]: I'm gonna trust you on this

**Review by pszkup7bulls — APPROVED** [review]

**Comment by LSZ-PL** [comment]
Hello Daniel,
please review my PR containing pricing functionality (pricing discovery + persistence + querying) and processing of first cloud service provider - AWS.

Thanks,
ŁSz

**Pull request merged**